ecosystem/typescript/sdk/src/plugins/ans_client.ts

```diff
@@ -0,0 +1,166 @@
+import { Provider } from "../providers";
+import * as Gen from "../generated/index";
+
+const ansContractsMap: Record<string, string> = {
+  testnet: "0x5f8fd2347449685cf41d4db97926ec3a096eaf381332be4f1318ad4d16a8497c",
+  mainnet: "0x867ed1f6bf916171b1de3ee92849b8978b7d1b9e0a8cc982a3d19d535dfd9c0c",
+};
+
+// Each name component can only have lowercase letters, number or hyphens, and cannot start or end with a hyphen.
+const nameComponentPattern = /^[a-z\d][a-z\d-]{1,61}[a-z\d]$/;
+
+const namePattern = new RegExp(
+  "^" +
+    // Optional subdomain (cannot be followed by .apt)
+    "(?:(?<subdomain>[^.]+)\\.(?!apt$))?" +
+    // Domain
+    "(?<domain>[^.]+)" +
+    // Optional .apt suffix
+    "(?:\\.apt)?" +
+    "$",
+);
+
+type ReverseLookupRegistryV1 = {
+  registry: {
+    handle: string;
+  };
+};
+
+type NameRegistryV1 = {
+  registry: {
+    handle: string;
+  };
+};
+
+export class AnsClient {
+  contractAddress: string;
+
+  provider: Provider;
+
+  /**
+   * Creates new AnsClient instance
+   * @param provider Provider instance
+   * @param contractAddress An optional contract address.
+   * If there is no contract address matching to the provided network
+   * then the AnsClient class expects a contract address -
+   * this is to support both mainnet/testnet networks and local development.
+   */
+  constructor(provider: Provider, contractAddress?: string) {
+    this.provider = provider;
+    if (!ansContractsMap[this.provider.network] && !contractAddress) {
+      throw new Error("Please provide a valid contract address");
+    }
+    this.contractAddress = ansContractsMap[this.provider.network] ?? contractAddress;
+  }
+
+  /**
+   * Returns the primary name for the given account address
+   * @param address An account address
+   * @returns Account's primary name | null
+   */
+  async getPrimaryNamebyAddress(address: string): Promise<string | null> {
+    const ansResource: Gen.MoveResource = await this.provider.getAccountResource(
+      this.contractAddress,
+      `${this.contractAddress}::domains::ReverseLookupRegistryV1`,
+    );
+    const data = ansResource.data as ReverseLookupRegistryV1;
+    const { handle } = data.registry;
+    const domainsTableItemRequest = {
+      key_type: "address",
+      value_type: `${this.contractAddress}::domains::NameRecordKeyV1`,
+      key: address,
+    };
+    try {
+      const item = await this.provider.getTableItem(handle, domainsTableItemRequest);
+      return item.subdomain_name.vec[0] ? `${item.subdomain_name.vec[0]}.${item.domain_name}` : item.domain_name;
+    } catch (error: any) {
```

Comment:
what if we get a different error, such as service not available?
is there a logging service available?

Reply:
No logging library in the TS SDK I believe, what kind of logging did you have in mind? Like push analytics?

Reply:
I was thinking on something like https://tslog.js.org/#/ - but anyway, https://github.com/aptos-labs/aptos-core/pull/7227#discussion_r1148045737